slateratom/lib/confinement.F90

```diff
@@ -0,0 +1,600 @@
+#:include 'common.fypp'
+
+!> Module that builds up various supervectors.
+module confinement
+
+  use common_accuracy, only : dp
+  use utilities, only : fak
+  use core_overlap, only : v
+  use density, only : basis_times_basis_times_r2
+
+  implicit none
+  private
+
+  public :: TConf, TConfInp, confType
+  public :: TPowerConf, TPowerConf_init
+  public :: TWsConf, TWsConf_init
+
+
+  !> Generic wrapper for confinement potentials.
+  type, abstract :: TConf
+  contains
+
+    procedure(getPotOnGrid), deferred :: getPotOnGrid
+    procedure(getSupervec), deferred :: getSupervec
```

Comment:
Do we really need `getSupervec()`? Both implementations are exact identical, if I have seen correctly. Let's introduce it only, if we really need it. Or, alternatively, create a concrete implementation in this base class, so that derived classes have only to define `getPotOnGrid()`

Reply:
Exactly, I was trying to sharpen my Fortran OOP skills with this PR and would be very interested in your opinion/feedback. The `getSupervec()` routine is a 1:1 duplicate and does not depend on `this`. I wasn't quite sure how to handle this case in the most elegant way. Since it does not depend on the compression type, can it be associated with the overarching abstract type (is this what you meant by 'create a concrete implementation in this base class'?) or should it just live on its own in the confinement module?

Reply:
* An abstract type can have non-deferred (normal) type bound procedures with actual implementation.
* You could either specify the routine there. You might then use the `nopass` option, so that `this` is not even passed to the routine. 
* Alternatively, you just use it as a simple function outside. I don't have any strong opinions on this, but probably the way, how the integrals are calculated is independent from the actual potential form, so probably having it outside as a standalone function is more logical...

Reply:
I see :+1:. I have seen `nopass` in other projects before and always wanted to use it, seems like a reasonable opportunity for that. See the latest commit.